Add jax.devices() and friends, and add `devices` arg to pmap.

skye opened the pull request:
This change adds the following APIs:
* jax.devices(). This returns a list of available Device subclass instances.
* jax.host_id(). Currently always 0, but will be useful on multi-host platforms.
* jax.local_device_count(). Currently always equal to jax.device_count(), but
    will be useful on multi-host platforms.
* Optional `devices` argument to pmap. This can be used to specify which devices
    should be used in the replicated computation.

mattjj commented on jax/api.py:
```diff
@@ -640,13 +642,19 @@ def pmap(fun, axis_name=None):
     fun: Function to be mapped over argument axes.
     axis_name: Optional, a hashable Python object used to identify the mapped
       axis so that parallel collectives can be applied.
+    devices: Optional, a sequence of Devices to map over. (Available devices can
```
Should we indicate this as an experimental API likely to change (even if it's not likely to change, just to provide some cover)?

mattjj commented on jax/api.py:
```diff
@@ -640,13 +642,19 @@ def pmap(fun, axis_name=None):
   The mapped axis size must be less than or equal to the number of XLA devices
-  available. For nested ``pmap`` calls, the product of the mapped axis sizes
-  must be less than or equal to the number of XLA devices.
+  available (unless `devices` is specified, see below). For nested ``pmap``
```
Maybe double-backtick around "devices" (for Sphinx formatting).

mattjj commented on jax/api.py:
```diff
@@ -707,6 +715,25 @@ def pmap(fun, axis_name=None):
   >>> print(doubly_normed.sum((0, 1)))
   1.0
+
+  The ``devices`` argument can be used to specify exactly which devices are used
```
Not blocking for this PR, but we could add a description of how the order matters, i.e. that it is significant and corresponds to the order along the mapped axis.

Could also specify what happens on nesting (error, ideally?).

mattjj commented on jax/interpreters/pxla.py:
```diff
@@ -159,17 +159,30 @@ def replica_groups(nrep, mesh_spec, mesh_axes):
+      raise ValueError(msg.format(num_replicas, xb.device_count()))
+    device_assignment = None
+  else:
+    assert all(d.host_id == xb.host_id() for d in devices)
```
Since this is checking for an error in a user-supplied argument, eventually it should probably be a ValueError rather than an assertion failure, but not blocking for this PR; just noting it for discussion and/or future ref!

mattjj commented on jax/interpreters/pxla.py:
```diff
@@ -394,13 +407,19 @@ def __getitem__(self, idx):
+  devices = params.pop('devices')
   assert not params
+
+  if devices is not None and len(devices) == 0:
```
This particular check is super cheap, but I recently got the idea from Peter that we could do all our error checking under the cached functions (i.e. under `parallel_callable` here), since that means we only do the error checking once (at "compile time") and not on each dispatch.

mattjj reviewed (approved):
This is fantastic.

None of my comments are substantial, and all are just suggestions that you should be ready and willing to ignore if you don't like them! I mainly wanted to show I read the code :)

This is really nice, though, including the minor cleanups on existing code.

skye reviewed (commented):
Thanks for the thorough review!

skye committed:
Add jax.devices() and friends, and add `devices` arg to pmap.

This change adds the following APIs:
* jax.devices(). This returns a list of available Device subclass instances.
* jax.host_id(). Currently always 0, but will be useful on multi-host platforms.
* jax.local_device_count(). Currently always equal to jax.device_count(), but
    will be useful on multi-host platforms.
* Optional `devices` argument to pmap. This can be used to specify which devices
    should be used in the replicated computation.
